node/src/components/linear_chain.rs

```diff
@@ -203,3 +210,445 @@ where
+            fixture.create_next_block(
+                ERA_3,
+                Some((era_4_validators.clone(), original_era_5_validators.clone())),
+                PublicKey::System,
+                VERSION_2,
+                era_3_validators,
+            );
+
+            // Block 8: non-switch block in era 4.
+            fixture.create_next_block(
+                ERA_4,
+                None,
+                era_4_validators.keys().next().unwrap().clone(),
+                VERSION_2,
+                era_4_validators.clone(),
+            );
+
+            // Block 9: switch block in era 4.
+            fixture.create_next_block(
+                ERA_4,
+                Some((original_era_5_validators.clone(), original_era_5_validators)),
+                era_4_validators.keys().next().unwrap().clone(),
+                VERSION_2,
+                era_4_validators,
+            );
+
+            // Block 10: immediate switch block in era 5 after validator-change upgrade to version
+            // 3.0.0.
+            fixture.create_next_block(
+                ERA_5,
+                Some((era_6_validators.clone(), era_7_validators)),
+                PublicKey::System,
+                VERSION_3,
+                replacement_era_5_validators,
+            );
+
+            // Block 11: non-switch block in era 6.
+            fixture.create_next_block(
+                ERA_6,
+                None,
+                era_6_validators.keys().next().unwrap().clone(),
+                VERSION_3,
+                era_6_validators,
+            );
+
+            fixture
+        }
+
+        fn create_next_block(
+            &mut self,
+            era: EraId,
+            next_two_eras_validator_weights: Option<(ValidatorWeights, ValidatorWeights)>,
+            proposer: PublicKey,
+            protocol_version: ProtocolVersion,
+            validators_for_this_block: ValidatorWeights,
+        ) {
+            let last_block = self.blocks.values().last().unwrap();
+            let era_report = next_two_eras_validator_weights
+                .as_ref()
+                .map(|_| EraReport::default());
+            let finalized_block = FinalizedBlock::new(
+                BlockPayload::default(),
+                era_report,
+                Timestamp::now(),
+                era,
+                last_block.height() + 1,
+                proposer,
+            );
+
+            let state_root_hash = Digest::hash(&[era.value() as u8]);
+            // If this is a switch block, update the validator weights to have entries for this era
+            // and the next two eras.  Otherwise, just copy the validator weights for the previous
+            // block.
+            let validators_in_global_state = match next_two_eras_validator_weights.as_ref() {
+                Some((weights_1, weights_2)) => {
+                    let mut validators = EraValidators::new();
+                    validators.insert(era, validators_for_this_block.clone());
```

Comment:
If this is a switch block, does the global state still contain this entry?

Reply:
I think actual switch blocks don't contain the era validators for themselves, only starting at the next era. Not 100% sure, though.

Reply:
Yeah - I think you're right.  I'll remove them.